src/Nest/Aggregations/AggregateJsonConverter.cs

```diff
@@ -10,580 +10,102 @@ namespace Nest
-						Value = value.GetValueOrDefault(0)
-					});
-					reader.Read();
-				}
-			}
-			metric.Items = percentileItems;
-			if (!oldFormat) reader.Read();
-			return metric;
-		}
-
-		private IAggregate GetSingleBucketAggregate(JsonReader reader, JsonSerializer serializer)
+		private IAggregate ReadAggregate(JObject jObject, JsonSerializer serializer)
 		{
-			reader.Read();
-			var docCount = (reader.Value as long?).GetValueOrDefault(0);
-			var bucket = new SingleBucketAggregate { DocCount = docCount };
-			reader.Read();
-			if (reader.TokenType == JsonToken.PropertyName && (string)reader.Value == "buckets")
-			{
-				var b = this.GetMultiBucketAggregate(reader, serializer) as BucketAggregate;
-				return new BucketAggregate
-				{
-					DocCount = docCount,
-					Items = b.Items
-				};
-			}
-
-			bucket.Aggregations = this.GetSubAggregates(reader, serializer);
-
-			return bucket;
-		}
-
-		private IAggregate GetStatsAggregate(JsonReader reader, JsonSerializer serializer)
-		{
-			reader.Read();
-			var count = (reader.Value as long?).GetValueOrDefault(0);
-			reader.Read(); reader.Read();
-			var min = (reader.Value as double?);
-			reader.Read(); reader.Read();
-			var max = (reader.Value as double?);
-			reader.Read(); reader.Read();
-			var average = (reader.Value as double?);
-			reader.Read(); reader.Read();
-			var sum = (reader.Value as double?);
-
-			var statsMetric = new StatsAggregate()
-			{
-				Average = average,
-				Count = count,
-				Max = max,
-				Min = min,
-				Sum = sum
-			};
-
-			reader.Read();
-
-			if (reader.TokenType == JsonToken.EndObject)
-				return statsMetric;
-
-			while (reader.TokenType != JsonToken.EndObject && (reader.Value as string).Contains("_as_string"))
-			{
-				reader.Read();
-				reader.Read();
-			}
+			var metaProperty = jObject.Property("meta");
+			if (metaProperty == null)
+				throw new Exception("Cannot deserialization aggregation result. The response is missing metadata.");
 
-			if (reader.TokenType == JsonToken.EndObject)
-				return statsMetric;
+			var meta = metaProperty.Value.ToObject<Dictionary<string, object>>();
+			if (!meta.ContainsKey(AggregationMetadata.Key))
+				throw new Exception($"Cannot deserialization aggregation result. Metadata is missing {AggregationMetadata.Key}.");
 
-			return GetExtendedStatsAggregate(statsMetric, reader);
-		}
+			var type = meta[AggregationMetadata.Key] as string;
+			// Remove the injected metadata so that we don't dirty the users results
+			meta.Remove(AggregationMetadata.Key);
 
-		private IAggregate GetExtendedStatsAggregate(StatsAggregate statsMetric, JsonReader reader)
-		{
-			var extendedStatsMetric = new ExtendedStatsAggregate()
-			{
-				Average = statsMetric.Average,
-				Count = statsMetric.Count,
-				Max = statsMetric.Max,
-				Min = statsMetric.Min,
-				Sum = statsMetric.Sum
-			};
+			var aggregate = jObject.ToObject(AggregationMetadata.Map[type]) as IAggregate;
+			aggregate.Meta = meta.HasAny() ? meta : null;
```

Comment:
No, because if our injected `_type` was the only value in the metadata object, then after we remove it will be assigning `Meta` to an empty dictionary, where it should be null instead.